### ISSUE TITLE
Add a ':write" command

### ISSUE THREAD

**Issue by TomMD** [GaloisInc/cryptol#216]
The CLI could use a `:write` command morally in the form `write : [a][8] -> FilePath -> IO ()` (or flipped).   It is getting tedious, copy-pasting 8KB of text into a Haskell program that then writes the hex into a binary file... and that's after only doing it once.

**Comment by acfoltzer** [comment]
Can you say more about the use case? I've thought about something similar before but for input, rather than output (say you want to have a _really_ big key, but don't want to force the parser to read it all in).

Also, does the type signature imply that you'd want it to just be raw binary, or ASCII chars?

**Comment by TomMD** [comment]
@acfoltzer Sure.  I was working on a file encapsulation format and even a bare hand-full of times is too much for copy and pasting an expression into GHCi to perform `writeFile . pack $ someCryptolOutput`.  So the idea, as implemented in patch ab54e71b052ac30ae9efa33579f00a30c2315075, is that the user could:

```
let v = someCryptolExpression
:set outputFile=/tmp/junkfile
:write v
```

There are a few short-comings and warts currently (some we might not care to tackle):

- No `:read` support
- Inconsistent names (`outputFile` vs `smtfile` notice the camel case)
- No stripping of quotes (`:set outputFile="/tmp/ThisDoesNotWork"`)
- I have no idea what I'm doing with Cryptol ASTs - are my assumptions valid? Do `[n][8]` types always come in the form of VSeq-VSeq-BV?  Can the type be in a different "non-normalized" form my hacky check won't recognize?  It seems possible.

**Comment by TomMD** [comment]
As a matter of update:

* You can observe `VSeq [VSeq [VBit ..` but a `pack` function exists to tranform the inner seq to `VWord`.
* There is no resistance to adding n-ary commands (`:write FileName Expr`), it was not preexisting but don't take that as an intentional design decision.

**Comment by acfoltzer** [comment]
> No :read support

Definitely want this eventually, but we should probably be careful about what formats we adopt here. Are there any standardized formats for blobs of bytes we could use, other than just the blob of bytes itself?

> Inconsistent names (outputFile vs smtfile notice the camel case)

I've been meaning to do a cleanup pass of these anyway. The inconsistency is not intended.

> No stripping of quotes (:set outputFile="/tmp/ThisDoesNotWork")

We should probably support this better, particularly for those users who might dare to use paths with spaces in them :wink:

> I have no idea what I'm doing with Cryptol ASTs - are my assumptions valid? Do [n][8] types always come in the form of VSeq-VSeq-BV? Can the type be in a different "non-normalized" form my hacky check won't recognize? It seems possible.
> You can observe VSeq [VSeq [VBit .. but a pack function exists to tranform the inner seq to VWord.

I'm not entirely clear on the invariants here either. Some of the functions involved in evaluation will turn the seq(seq(bit)) case into a seq(bv), but we recently had to make a change backing off of that because it can introduce extra strictness. @brianhuffman might be able to illuminate further.

> There is no resistance to adding n-ary commands (:write FileName Expr), it was not preexisting but don't take that as an intentional design decision.

It is merely preexisting, not intentional. I think that rather than letting >0-ary commands proliferate with accompanying `fooFile=` settings, we should just implement n-ary commands properly.

I'll check out the diff in a bit.

**Comment by TomMD** [comment]
Ah, I should have updated this ticket a little after talking to yav yesterday:

- There is now read support as part of the pull request (#220), see 8776f2c854aa28c0e9339c52aefa9df1ded3fa0b

- I eliminated the use of "outputFile" but there are still consistency issues. For example, the very use of "smtFile" I think should actually be removed and a new write-like command such as `:writeSMT filepath expr` should be added.  Camel-case vs underscores debatable.

- I _think_ I have the invariants right now thanks to @brianhuffman  and @yav.  For the expression I use `fromVWord` to handle the only two possible cases of `VSeq False [VWord]` and `VSeq False [VSeq True _]`.  For the type I leveraged pre-existing functions `tIsSeq`, `tIsBit` and `tIsNum` - if `[n][8]` can appear in a form other than `TSeq [TSeq TBit | TNum {width=8}]` then that would be good to know.

- I have added a 2-ary command with the constructor `FileExprArg` and dropped the extra environmental variable.  The implementation involves some string munging that *will fail in the case of file paths with spaces* (notable since you mentioned allowing quoted filenames for this reason).

**Comment by acfoltzer** [comment]
Excellent, yeah, I wish I'd looked at the diff first :wink:

I think we ought to just add a few test cases (including a filenames-with-spaces test), merge, and then refine. This will be a nice forcing function for improvements on other interface issues.

**Comment by TomMD** [comment]
Closed dup of #220
